### PR TITLE
Update ConfigMap EscalationPolicyID data field on reconcile

### PR DESCRIPTION
**Pull request by supreeth7**
Relates to: [OSD-12385](https://issues.redhat.com/browse/OSD-12385)
<hr>

- Currently, for existing PagerDuty services, the PD ConfigMaps do not have the `EscalationPolicyID` data field. Only the newly created service ConfigMaps have it.
- This is necessary to detect when there is a change in the PagerDutyIntegration escalation policy, compare it with the existing ConfigMap escalation policy, update the service and reconcile.

#### Changes with this PR
- Default to using PagerDutyIntegration CR's `EscalationPolicyID` as the source of truth for PagerDuty Operator escalation policy. This makes the `PDIEscalationPolicyID` field obsolete.
- Check if the ClusterDeployment's ConfigMap has the `ESCALATION_POLICY_ID` key, if not, map it to the PagerDutyIntegration CR `EscalationPolicyID` and update the ConfigMap. This will change the PagerDuty ConfigMap for all existing ClusterDeployments.
- Compare the ConfigMap escalation policy ID with the PDI escalation policy, if there's a change detected, update the PagerDuty service, and update the relevant ConfigMap. This is true only if the ConfigMap has the `ESCALATION_POLICY_ID` key and it is not empty.
- Implement unit tests to verify the above changes.

### CODE REVIEW

**Review by jewzaam — CHANGES_REQUESTED** [review]
/hold

PDIEscalationPolicyID vs EscalationPolicyID in Data struct is very hard to follow.
This PR changes EscalationPolicyID to be loaded from the CM and is used to check if there needs to be an update.  Furthermore, on CM load the PDIEscalationPolicyID is not set.  Any CM that has missing ESCALATION_POLICY_ID key will never have PDIEscalationPolicyID set and therefore will never get ESCALATION_POLICY_ID set.

1. Consolidate on use of single EscalationPolicyID as it is not possible to know what is happening without tracking through a lot of logic.
2. Add unit tests to cover the bug you are fixing.  Make sure the test reproduces the bug before you fix it.
3. Add unit tests to cover additional cases that might make sense for missing or unexpected values.

**Comment by supreeth7** [comment]
I agree, even though I was having a hard time trying to differentiate and track the two separate escalation policies.
I have made the appropriate changes and updated the description to reflect the same.
@jewzaam can you please give this another look? Thanks.

File: pkg/pagerduty/service.go [hunk 1]
```diff
@@ -175,15 +174,6 @@ func (data *Data) ParseClusterConfig(osc client.Client, namespace string, cmName
...
 
-	data.EscalationPolicyID, err = getConfigMapKey(pdAPIConfigMap.Data, "ESCALATION_POLICY_ID")
-	if err != nil {
-		// If data.PDIEscalationPolicyID exists, use that as a default
-		if data.PDIEscalationPolicyID == "" {
-			return err
-		}
-		data.EscalationPolicyID = data.PDIEscalationPolicyID
-	}
-
```
[the remark this answers is not in the document]
Comment (supreeth7) [reply pkg/pagerduty/service.go]: I left the aspect of code modularity out of the window on this one, thanks for bringing it up.

**Comment by supreeth7** [comment]
Did a sanity check on a staging cluster, with the changes in https://github.com/supreeth7/pagerduty-operator/pull/1 merged, everything is working as expected.

**Comment by jewzaam** [comment]
Thank you @supreeth7, looks great.

/hold cancel
/lgtm
